examples/tfengine/generated/org_foundation/cicd/triggers/validate/main.tf

```diff
@@ -0,0 +1,102 @@
+# Copyright 2021 Google LLC
```

Comment:
I still think the 3 trigger submodules are too similar and we can reuse and just have 1 submodule for one trigger. Let's talk about it tomorrow.

Reply:
So cicd/main.tf will call cicd/triggers/main.tf which will call cicd/triggers/trigger/main.tf 3 times. Uncommon part between different triggers such as `"${var.terraform_root_prefix}cicd/configs/tf-apply.yaml"` can be passed in as a variable from one level up, i.e. from cicd/triggers/main.tf

in cicd/triggers/main.tf, it will be like

```
module "validate_triggers" {
  source = "./validate"

  branch_name           = var.branch_name
  managed_dirs          = var.managed_dirs
  name                  = var.name
  skip                  = var.triggers.validate.skip
  run_on_push           = var.triggers.validate.run_on_push
  run_on_schedule       = var.triggers.validate.run_on_schedule
  github                = var.github
  project_id            = var.project_id
  scheduler_region      = var.scheduler_region
  terraform_root        = var.terraform_root
  terraform_root_prefix = var.terraform_root_prefix
  build_config_file = "${var.terraform_root_prefix}cicd/configs/tf-validate.yaml"
  ... other vars similar to build_config_file
}

module "plan_triggers" {
  source = "./plan"

  branch_name           = var.branch_name
  managed_dirs          = var.managed_dirs
  name                  = var.name
  skip                  = var.triggers.plan.skip
  run_on_push           = var.triggers.plan.run_on_push
  run_on_schedule       = var.triggers.plan.run_on_schedule
  github                = var.github
  project_id            = var.project_id
  scheduler_region      = var.scheduler_region
  terraform_root        = var.terraform_root
  terraform_root_prefix = var.terraform_root_prefix
  build_config_file = "${var.terraform_root_prefix}cicd/configs/tf-plan.yaml"
}

# one more for apply
...
...
```

Reply:
So I actually looked at this for about half an hour and I couldn't find a good way to conditionally select between configuring github and CSR. We'd have to have something like a ./csr_trigger and ./github_trigger and call those based on template vars in ./triggers/main.tf.

Reply:
I pushed up a branch with roughly what I'm talking about (haven't tested it): https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/tree/single_trigger_submodule/templates/tfengine/components/cicd/triggers